### ISSUE TITLE
cron-{daily, weekly, monthly}.service (and boot_delay) is part of the system's startup time

### ISSUE THREAD

**Issue by a-detiste** [systemd-cron/systemd-cron#50]
Forwarded from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=829438 .

I have seen the same behaviour, but I have no clue.

----------------

Systemd service cron-daily.service has line

    ExecStartPre=/lib/systemd-cron/boot_delay 5

which makes it wait until system's uptime is at least 5 minutes. The
delay is also part of Systemd's startup time as reported by journal:

    [...] systemd[1]: Startup finished in 4.159s (kernel) + 4min 58.016s (userspace) = 5min 2.175s.

See the userspace and total time. The delay also shows in the graph
produced by command "systemd-analyze plot" (I'll try to include an
attachment).

This is a minor issue but it would be nice if cron services weren't part
of the startup time.

Teemu Likonen

**Comment by a-detiste** [comment]
`systemd-analyze critical-chain` can be usefull to debug

**Comment by a-detiste** [comment]
https://www.linux-magazine.com/Issues/2018/212/The-sys-admin-s-daily-grind-systemd-analyze

**Comment by hifron** [comment]
SystemD services has support to start after some occurrence or event happens. Like graphical start, network online or maybe some boot level(coming from the SysV init era run levels), but SystemD is not everywhere like Devuan in unix Space like to remind...

```
[Unit]
Description=Some service to start
After=graphical-session.target network-online.target dconf.service

[Service]
Type=oneshot
ExecStart=/some_path/...

[Install]
WantedBy=graphical-session.target
```

so if systemd-cron requires something from systemd, this should be initialized and systemd programmers should know what to do.

[1] _In user login there is need for `dbus-update-activation-environment --systemd --all` in .profile but that's not guaranteed for apps support like gnome-clocks which is known bug..._

[1] arch wiki dbus

**Comment by nabijaczleweli** [comment]
Am I the only one who doesn't understand what this post is saying?

**Comment by a-detiste** [comment]
It looks like AI nonsense

**Comment by isodude** [comment]
Isn't the solution here to add
```
[Timer]
OnBootSec=5m
```
to the cron-daily.timer?

**Comment by nabijaczleweli** [comment]
Don't think so, that would schedule the daily job at midnight *and also* at boot+5min right?

**Comment by isodude** [comment]
You are correct. I did see that [boot-complete.target](https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html#boot-complete.target) exists, but that only works for UEFI-systems. What is the reason for the boot-delay-script to start with?

**Comment by nabijaczleweli** [comment]
cron-boot.service is OnBootSec=60,
the /etc/cron.${hourly,daily,whatever} ones mimick this by generating with
increasing (uptime-5min)/(uptime-10min)/&c. delays.

The reason is not explicitly documented anywhere,
but it follows to me that you naturally want to stagger these anyway,
with respect to cron-boot, each other, and the actual system boot,
and you want to run them on a mostly-settled-down system.

**Comment by isodude** [comment]
Would it make sense to add `After=cron-boot.service` to daily and `After=cron-daily.service` to hourly etc.. ?
Or a more convoluted way, adding `cron-delay@300.timer` with `cron-delay@300.service` and then `After=cron-delay@300.service` in cron-daily.service. Also nothing stopping to do both really (to force staggering).
I guess `Type=idle` could help out somewhat to make sure that it waits for other units to finish.

**Comment by a-detiste** [comment]
The `runparts` mode is the historic, original way that cron/systemd-cron worked. For mort usefull logging/finer control, the default setting for systemd-cron is now to generate one .service/.timer pair per item in /etc/cron.{hourly|daily|weekly|...}/

It makes more sense nowadays because a lot of package providing these items (at least on Debian-based distro) will also provide a native .service and the old crontab would only check for running under systemd & exit immediatly.

That all said, it's the reason I won't invest much more into the old runparts mode (I don't use anymore). It might still be a better fit for Redhat distro users that are used to the sequencing provided by runparts.

**Comment by a-detiste** [comment]
> The reason is not explicitly documented anywhere,

It mimicks what is in the /etc/crontab provided by `cron-daemon-common`;
and what people usually exepcts from maintenance tasks that should not slow down the boot.

**Comment by isodude** [comment]
I don't see how separate timers solve staggering/boot_delay though.

**Comment by nabijaczleweli** [comment]
I don't think those `After=`s make much sense (I'm especially not thrilled about the interactions between the services). Adding a Wants=/After= for a ExecStart=/bin/true,Persistent=1 cron-bootdelay@.service/.timer, while it may in fact work, is also distinctly mid I think, especially for an issue that appears entirely cosmetic.
